README.md
# linear_step_indicator

A Flutter package for adding beautiful step indicators in your apps.

## Install

In the `pubspec.yaml` of your flutter project, add the following dependency:

```yaml
dependencies:
  linear_step_indicator: ^1.0.0+3
```

Import the package in your project:

```dart
import 'package:linear_step_indicator/linear_step_indicator.dart';
```

## Getting Started

Examples:

```dart
FullLinearStepIndicator(
          steps: 5,
          lineHeight: 3.5,
          activeNodeColor: Colors.brown,
          inActiveNodeColor: const Color(0xffd1d5d8),
          activeLineColor: Colors.brown,
          inActiveLineColor: const Color(0xffd1d5d8),
          controller: pageController,
          labels: List<String>.generate(STEPS, (index) => "Step ${index + 1}"),
          complete: () {
            //typically, you'd want to put logic that returns true when all the steps
            //are completed here
            return Future.value(true);
          },
        )
```



![Indicator](https://github.com/Crazelu/linear-step-indicator/blob/main/assets/full-linear-indicator.png)


```dart
LinearStepIndicator(
          steps: 3,
          controller: _pageController,
          labels: List<String>.generate(3, (index) => "Step ${index + 1}"),
          complete: () {
            //typically, you'd want to put logic that returns true when all the steps
            //are completed here
            return Future.value(true);
          },
        )
```



![Indicator](https://github.com/Crazelu/linear-step-indicator/blob/main/assets/indicator.png)




```dart
StepIndicatorPageView(
        steps: 3,
        indicatorPosition: IndicatorPosition.top,
        labels: List<String>.generate(3, (index) => "Step ${index + 1}"),
        controller: _pageController,
        complete: () {
          //typically, you'd want to put logic that returns true when all the steps
          //are completed here
          return Future.value(true);
        },
        children: List<Widget>.generate(
          3,
          (index) => Container(
            color: Color(0xffffffff),
            child: Center(
              child: Text(
                "Page ${index + 1}",
                style: TextStyle(
                  fontSize: 24,
                  fontWeight: FontWeight.w500,
                ),
              ),
            ),
          ),
        ),
      )
```

<img src="https://raw.githubusercontent.com/Crazelu/linear-step-indicator/main/assets/indicator-page.png" width="400" height="800">

## Contributions

Feel free to contribute to this project.

If you find a bug or want a feature, but don't know how to fix/implement it, please fill an [issue](https://github.com/Crazelu/linear-step-indicator/issues).  
If you fixed a bug or implemented a feature, please send a [pull request](https://github.com/Crazelu/linear-step-indicator/pulls).


CHANGELOG.md
## 1.0.0+3

* Adds FullLinearStepIndicator
* Adds support for scroll control on StepIndicatorPageView with ScrollPhysics

## 1.0.0+2

* Removes explicit Flutter version

## 1.0.0+1

* Updates image assets in README
* Adjusts alignment of label texts
## 1.0.0

* Initial release.


test/linear_step_indicator_test.dart
// import 'package:flutter_test/flutter_test.dart';

// import 'package:linear_step_indicator/linear_step_indicator.dart';

// void main() {
//   test('adds one to input values', () {
//     final calculator = Calculator();
//     expect(calculator.addOne(2), 3);
//     expect(calculator.addOne(-7), -6);
//     expect(calculator.addOne(0), 1);
//   });
// }


lib/src/step_indicator_page_view.dart
import 'package:flutter/material.dart';
import 'package:linear_step_indicator/src/linear_step_indicator.dart';
import 'package:linear_step_indicator/src/constants.dart';

class StepIndicatorPageView extends StatelessWidget {
  ///Controller for tracking page changes.
  ///
  ///Typically, controller should animate or jump to next page
  ///when a step is completed
  final PageController controller;

  ///Number of nodes to paint on screen
  final int steps;

  ///StepIndicatorPageView's children widgets
  final List<Widget> children;

  ///[completedIcon] size
  final double iconSize;

  ///Size of each node
  final double nodeSize;

  ///Height of separating line
  final double lineHeight;

  ///Icon showed when a step is completed
  final IconData completedIcon;

  ///Color of each completed node border
  final Color activeBorderColor;

  ///Color of each uncompleted node border
  final Color inActiveBorderColor;

  ///Color of each separating line after a completed node
  final Color activeLineColor;

  ///Color of each separating line after an uncompleted node
  final Color inActiveLineColor;

  ///Background color of a completed node
  final Color activeNodeColor;

  ///Background color of an uncompleted node
  final Color inActiveNodeColor;

  ///Thickness of node's borders
  final double nodeThickness;

  ///Node's shape
  final BoxShape shape;

  ///[completedIcon] color
  final Color iconColor;

  ///Step indicator's background color
  final Color backgroundColor;

  ///Boolean function that returns [true] when last node should be completed
  final Complete? complete;

  ///Step indicator's vertical padding
  final double verticalPadding;

  ///Spacing from top of the screen if [indicatorPosition] is [IndicatorPosition.top]
  final double spacing;

  ///Spacing from bottom of the screen if [indicatorPosition] is [IndicatorPosition.bottom]
  final double bottomSpacing;

  ///Indicator's position
  final IndicatorPosition indicatorPosition;

  ///Labels for individual nodes
  final List<String> labels;

  ///Textstyle for an active label
  final TextStyle? activeLabelStyle;

  ///Textstyle for an inactive label
  final TextStyle? inActiveLabelStyle;

  /// How the page view should respond to user input.
  ///
  /// For example, determines how the page view continues to animate after the
  /// user stops dragging the page view.
  ///
  /// The physics are modified to snap to page boundaries using
  /// [PageScrollPhysics] prior to being used.
  ///
  /// Defaults to matching platform conventions.
  final ScrollPhysics? physics;

  const StepIndicatorPageView({
    Key? key,
    required this.steps,
    required this.controller,
    required this.children,
    this.activeBorderColor = kActiveColor,
    this.inActiveBorderColor = kInActiveColor,
    this.activeLineColor = kActiveLineColor,
    this.inActiveLineColor = kInActiveLineColor,
    this.activeNodeColor = kActiveColor,
    this.inActiveNodeColor = kInActiveNodeColor,
    this.iconSize = kIconSize,
    this.completedIcon = kCompletedIcon,
    this.nodeThickness = kDefaultThickness,
    this.nodeSize = kDefaultSize,
    this.verticalPadding = kDefaultSize,
    this.lineHeight = kDefaultLineHeight,
    this.shape = BoxShape.circle,
    this.iconColor = kIconColor,
    this.backgroundColor = kIconColor,
    this.spacing = kDefaultSize,
    this.bottomSpacing = 0,
    this.indicatorPosition = IndicatorPosition.top,
    this.labels = const <String>[],
    this.complete,
    this.activeLabelStyle,
    this.inActiveLabelStyle,
    this.physics,
  }) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return SafeArea(
      child: Container(
        color: backgroundColor,
        child: Column(
          children: [
            if (indicatorPosition == IndicatorPosition.top) ...[
              LinearStepIndicator(
                labels: labels,
                steps: steps,
                controller: controller,
                backgroundColor: backgroundColor,
                complete: complete,
                activeBorderColor: activeBorderColor,
                inActiveBorderColor: inActiveBorderColor,
                activeLineColor: activeLineColor,
                inActiveLineColor: inActiveLineColor,
                activeNodeColor: activeNodeColor,
                inActiveNodeColor: inActiveNodeColor,
                shape: shape,
                iconColor: iconColor,
                iconSize: iconSize,
                verticalPadding: verticalPadding,
                completedIcon: completedIcon,
                lineHeight: lineHeight,
                nodeSize: nodeSize,
                nodeThickness: nodeThickness,
                activeLabelStyle: activeLabelStyle,
                inActiveLabelStyle: inActiveLabelStyle,
              ),
              SizedBox(height: spacing),
            ],
            Expanded(
              child: PageView(
                physics: physics,
                controller: controller,
                children: children,
              ),
            ),
            if (indicatorPosition == IndicatorPosition.bottom) ...[
              SizedBox(height: spacing),
              LinearStepIndicator(
                steps: steps,
                controller: controller,
                backgroundColor: backgroundColor,
                complete: complete,
                activeBorderColor: activeBorderColor,
                inActiveBorderColor: inActiveBorderColor,
                activeLineColor: activeLineColor,
                inActiveLineColor: inActiveLineColor,
                activeNodeColor: activeNodeColor,
                inActiveNodeColor: inActiveNodeColor,
                shape: shape,
                iconColor: iconColor,
                iconSize: iconSize,
                verticalPadding: verticalPadding,
                completedIcon: completedIcon,
                lineHeight: lineHeight,
                nodeSize: nodeSize,
                nodeThickness: nodeThickness,
              ),
              SizedBox(height: bottomSpacing),
            ],
          ],
        ),
      ),
    );
  }
}

enum IndicatorPosition { top, bottom }


lib/src/full_linear_step_indicator.dart
import 'package:flutter/material.dart';
import 'package:linear_step_indicator/src/constants.dart';
import 'extensions.dart';
import 'node.dart';

class FullLinearStepIndicator extends StatefulWidget {
  ///Controller for tracking page changes.
  ///
  ///Typically, controller should animate or jump to next page
  ///when a step is completed
  final PageController controller;

  ///Number of nodes to paint on screen
  final int steps;

  ///[completedIcon] size
  final double iconSize;

  ///Size of each node
  final double nodeSize;

  ///Height of separating line
  final double lineHeight;

  ///Icon showed when a step is completed
  final IconData completedIcon;

  ///Color of each completed node border
  final Color activeBorderColor;

  ///Color of each uncompleted node border
  final Color inActiveBorderColor;

  ///Color of each separating line after a completed node
  final Color activeLineColor;

  ///Color of each separating line after an uncompleted node
  final Color inActiveLineColor;

  ///Background color of a completed node
  final Color activeNodeColor;

  ///Background color of an uncompleted node
  final Color inActiveNodeColor;

  ///Thickness of node's borders
  final double nodeThickness;

  ///Node's shape
  final BoxShape shape;

  ///[completedIcon] color
  final Color iconColor;

  ///Step indicator's background color
  final Color backgroundColor;

  ///Boolean function that returns [true] when last node should be completed
  final Complete? complete;

  ///Step indicator's vertical padding
  final double verticalPadding;

  ///Labels for individual nodes
  final List<String> labels;

  ///Textstyle for an active label
  final TextStyle? activeLabelStyle;

  ///Textstyle for an inactive label
  final TextStyle? inActiveLabelStyle;

  const FullLinearStepIndicator({
    Key? key,
    required this.steps,
    required this.controller,
    this.activeBorderColor = kActiveColor,
    this.inActiveBorderColor = kInActiveColor,
    this.activeLineColor = kActiveLineColor,
    this.inActiveLineColor = kInActiveLineColor,
    this.activeNodeColor = kActiveColor,
    this.inActiveNodeColor = kInActiveNodeColor,
    this.iconSize = kIconSize,
    this.completedIcon = kCompletedIcon,
    this.nodeThickness = kDefaultThickness,
    this.nodeSize = kDefaultSize,
    this.verticalPadding = kDefaultSize,
    this.lineHeight = kDefaultLineHeight,
    this.shape = BoxShape.circle,
    this.iconColor = kIconColor,
    this.backgroundColor = kIconColor,
    this.complete,
    this.labels = const <String>[],
    this.activeLabelStyle,
    this.inActiveLabelStyle,
  })  : assert(steps > 0, "steps value must be a non-zero positive integer"),
        assert(labels.length == steps || labels.length == 0,
            "Provide exactly $steps strings for labels"),
        super(key: key);

  @override
  _FullLinearStepIndicatorState createState() =>
      _FullLinearStepIndicatorState();
}

class _FullLinearStepIndicatorState extends State<FullLinearStepIndicator> {
  late List<Node> nodes;
  late int lastStep;

  @override
  void initState() {
    super.initState();
    nodes = List<Node>.generate(widget.steps, (index) => Node(step: index));
    lastStep = 0;

    //listen to page changes to track when each step is ideally completed

    widget.controller.addListener(
      () async {
        if (widget.controller.page! > lastStep) {
          setState(
            () {
              nodes[lastStep].completed = true;
              lastStep = widget.controller.page!.ceil();
            },
          );
        }

        //checks if the controller has hit the max step
        //and checks [complete] to complete last node (or not)

        if (widget.controller.page! == widget.steps - 1 &&
            widget.complete != null) {
          if (await widget.complete!()) {
            nodes[widget.steps - 1].completed = true;
            setState(() {});
          }
        }
      },
    );
  }

  @override
  Widget build(BuildContext context) {
    final borderSide = BorderSide(color: Colors.white, width: 2.5);
    return Material(
      child: Container(
        padding: EdgeInsets.symmetric(vertical: widget.verticalPadding),
        color: widget.backgroundColor,
        child: Column(
          mainAxisSize: MainAxisSize.min,
          children: [
            Row(
              mainAxisAlignment: MainAxisAlignment.center,
              children: [
                for (var node in nodes) ...[
                  if (nodes.indexOf(node) == 0) ...{
                    Container(
                      color: node.completed
                          ? widget.activeLineColor
                          : widget.inActiveLineColor,
                      height: widget.lineHeight,
                      width: context.screenWidth(1 / widget.steps) * .25,
                    ),
                  },
                  Container(
                    alignment: Alignment.center,
                    height: 38,
                    width: 38,
                    decoration: BoxDecoration(
                      shape: BoxShape.circle,
                      color: node.completed
                          ? widget.activeNodeColor
                          : widget.inActiveNodeColor,
                      border: Border(
                        top: borderSide,
                        bottom: borderSide,
                        left: borderSide,
                        right: borderSide,
                      ),
                      boxShadow: [
                        BoxShadow(
                          spreadRadius: .5,
                          blurRadius: .5,
                          color: Theme.of(context)
                              .primaryColorDark
                              .withOpacity(.4),
                        ),
                      ],
                    ),
                    child: Text(
                      "${nodes.indexOf(node) + 1}",
                      style: TextStyle(
                        color: Colors.white,
                      ),
                    ),
                  ),
                  if (node.step != widget.steps - 1)
                    Container(
                      color: node.completed
                          ? widget.activeLineColor
                          : widget.inActiveLineColor,
                      height: widget.lineHeight,
                      width: widget.steps > 3
                          ? context.screenWidth(1 / widget.steps) - 40
                          : context.screenWidth(1 / widget.steps) - 28,
                    ),
                  if (nodes.indexOf(node) == widget.steps - 1) ...{
                    Container(
                      color: node.completed
                          ? widget.activeLineColor
                          : widget.inActiveLineColor,
                      height: widget.lineHeight,
                      width: context.screenWidth(1 / widget.steps) * .25,
                    ),
                  },
                ],
              ],
            ),
            SizedBox(height: 6),
            if (widget.labels.length > 0) ...[
              Row(
                mainAxisSize: MainAxisSize.min,
                children: [
                  for (int i = 0; i < widget.labels.length; i++) ...[
                    Text(
                      widget.labels[i],
                      style: nodes[i].completed
                          ? widget.activeLabelStyle
                          : widget.inActiveLabelStyle,
                    ),
                    if (widget.labels[i] !=
                        widget.labels[widget.steps - 1]) ...[
                      SizedBox(
                        width: widget.steps > 3
                            ? context.screenWidth(1 / widget.steps) - 45
                            : context.screenWidth(1 / widget.steps) - 35,
                      ),
                    ],
                  ],
                ],
              ),
            ],
          ],
        ),
      ),
    );
  }
}


example/lib/main.dart
import 'dart:async';
import 'package:flutter/material.dart';
import 'package:linear_step_indicator/linear_step_indicator.dart';

const int STEPS = 5;

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      debugShowCheckedModeBanner: false,
      title: 'Step Indicator Demo',
      theme: ThemeData(
        primarySwatch: Colors.green,
      ),
      home: FullLinearIndicatorDemo(),
    );
  }
}

class FullLinearIndicatorDemo extends StatefulWidget {
  const FullLinearIndicatorDemo({Key? key}) : super(key: key);

  @override
  _FullLinearIndicatorDemoState createState() =>
      _FullLinearIndicatorDemoState();
}

class _FullLinearIndicatorDemoState extends State<FullLinearIndicatorDemo> {
  final pageController = PageController();
  int initialPage = 0;

  @override
  void initState() {
    super.initState();
    WidgetsBinding.instance?.addPostFrameCallback(
      (timeStamp) {
        Timer.periodic(
          const Duration(milliseconds: 350),
          (_) {
            if (mounted) {
              initialPage += 1;
              if (initialPage == STEPS - 1) {
              } else {
                pageController.jumpToPage(initialPage);
              }
            }
          },
        );
      },
    );
  }

  @override
  Widget build(BuildContext context) {
    return Container(
      color: Colors.grey,
      child: Center(
        child: FullLinearStepIndicator(
          steps: STEPS,
          lineHeight: 3.5,
          activeNodeColor: Colors.brown,
          inActiveNodeColor: const Color(0xffd1d5d8),
          activeLineColor: Colors.brown,
          inActiveLineColor: const Color(0xffd1d5d8),
          controller: pageController,
          labels: List<String>.generate(STEPS, (index) => "Step ${index + 1}"),
          complete: () {
            //typically, you'd want to put logic that returns true when all the steps
            //are completed here
            return Future.value(true);
          },
        ),
      ),
    );
  }
}

class StepIndicatorDemo extends StatelessWidget {
  const StepIndicatorDemo({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return Container(
      child: Center(
        child: LinearStepIndicator(
          steps: STEPS,
          controller: PageController(),
          labels: List<String>.generate(STEPS, (index) => "Step ${index + 1}"),
          complete: () {
            //typically, you'd want to put logic that returns true when all the steps
            //are completed here
            return Future.value(true);
          },
        ),
      ),
    );
  }
}

class StepIndicatorPageViewDemo extends StatefulWidget {
  const StepIndicatorPageViewDemo({Key? key}) : super(key: key);

  @override
  _StepIndicatorPageViewDemoState createState() =>
      _StepIndicatorPageViewDemoState();
}

class _StepIndicatorPageViewDemoState extends State<StepIndicatorPageViewDemo> {
  final _pageController = PageController();
  @override
  Widget build(BuildContext context) {
    return Material(
      child: StepIndicatorPageView(
        steps: STEPS,
        indicatorPosition: IndicatorPosition.top,
        labels: List<String>.generate(STEPS, (index) => "Step ${index + 1}"),
        controller: _pageController,
        complete: () {
          //typically, you'd want to put logic that returns true when all the steps
          //are completed here
          return Future.value(true);
        },
        children: List<Widget>.generate(
          STEPS,
          (index) => Container(
            color: Color(0xffffffff),
            child: Center(
              child: Text(
                "Page ${index + 1}",
                style: TextStyle(
                  fontSize: 24,
                  fontWeight: FontWeight.w500,
                ),
              ),
            ),
          ),
        ),
      ),
    );
  }
}
